glean-core/src/metrics/object.rs

```diff
@@ -0,0 +1,155 @@
+        }
+    }
+
+    /// Sets to the specified structure.
+    ///
+    /// # Arguments
+    ///
+    /// * `glean` - the Glean instance this metric belongs to.
+    /// * `value` - the value to set.
+    #[doc(hidden)]
+    pub fn set_sync(&self, glean: &Glean, value: JsonValue) {
+        let value = Metric::Object(serde_json::to_string(&value).unwrap());
+        glean.storage().record(glean, &self.meta, &value)
+    }
+
+    /// Sets to the specified structure.
+    ///
+    /// No additional verification is done.
+    /// The shape needs to be externally verified.
+    ///
+    /// # Arguments
+    ///
+    /// * `value` - the value to set.
+    pub fn set(&self, value: JsonValue) {
+        let metric = self.clone();
+        crate::launch_with_glean(move |glean| metric.set_sync(glean, value))
+    }
+
+    /// Record an `InvalidValue` error for this metric.
+    ///
+    /// Only to be used by the RLB.
+    // TODO(bug 1691073): This can probably go once we have a more generic mechanism to record
+    // errors
+    pub fn record_schema_error(&self) {
+        let metric = self.clone();
+        crate::launch_with_glean(move |glean| {
+            let msg = "Value did not match predefined schema";
+            record_error(glean, &metric.meta, ErrorType::InvalidValue, msg, None);
+        });
+    }
+
+    /// Sets to the specified structure from a string in JSON encoding
+    ///
+    /// # Arguments
+    ///
+    /// * `glean` - the Glean instance this metric belongs to.
+    /// * `value` - the value to set.
+    pub fn set_str(&self, value: String) {
+        let metric = self.clone();
+        crate::launch_with_glean(move |glean| {
+            let value = match serde_json::from_str(&value) {
+                Ok(s) => s,
+                Err(_) => {
+                    let msg = "Cannot set invalid json";
+                    record_error(glean, &metric.meta, ErrorType::InvalidValue, msg, None);
+                    return;
+                }
+            };
+            metric.set_sync(glean, value)
+        })
+    }
+
+    /// Get current value
+    #[doc(hidden)]
+    pub fn get_value<'a, S: Into<Option<&'a str>>>(
+        &self,
+        glean: &Glean,
+        ping_name: S,
+    ) -> Option<String> {
+        let queried_ping_name = ping_name
+            .into()
+            .unwrap_or_else(|| &self.meta().inner.send_in_pings[0]);
+
+        match StorageManager.snapshot_metric_for_test(
+            glean.storage(),
+            queried_ping_name,
+            &self.meta.identifier(glean),
+            self.meta.inner.lifetime,
+        ) {
+            Some(Metric::Object(o)) => Some(o),
+            _ => None,
+        }
+    }
+
+    /// **Test-only API (exported for FFI purposes).**
+    ///
+    /// Gets the currently stored value as JSON-encoded string.
+    ///
+    /// This doesn't clear the stored value.
+    pub fn test_get_value(&self, ping_name: Option<String>) -> Option<String> {
```

Comment:
Comment: It does strike me as... asymmetric that we set a `JsonValue` but get a `String`. I understand why get is a `String`: glean-core doesn't know the structure. And I'm guessing set is a `JsonValue` because it's the strongest type we can manage.

Reply:
yeah, to be addressed in a followup